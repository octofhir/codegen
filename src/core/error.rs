//! Error types for codegen

use thiserror::Error;

/// Result type alias
pub type Result<T> = std::result::Result<T, Error>;

/// Main error type for codegen operations
#[derive(Error, Debug)]
pub enum Error {
    /// IO error
    #[error("IO error: {0}")]
    Io(#[from] std::io::Error),

    /// JSON parsing error
    #[error("JSON error: {0}")]
    Json(#[from] serde_json::Error),

    /// TOML parsing error
    #[error("TOML error: {0}")]
    Toml(#[from] toml::de::Error),

    /// Canonical manager error
    #[error("Canonical manager error: {0}")]
    CanonicalManager(String),

    /// Parser error
    #[error("Parser error: {0}")]
    Parser(String),

    /// Generator error
    #[error("Generator error: {0}")]
    Generator(String),

    /// Validation error
    #[error("Validation error: {0}")]
    Validation(String),

    /// Configuration error
    #[error("Configuration error: {0}")]
    Config(String),

    /// Template error
    #[error("Template error: {0}")]
    Template(String),

    /// Other errors
    #[error("{0}")]
    Other(String),
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_error_types() {
        let err = Error::Parser("test error".to_string());
        assert_eq!(err.to_string(), "Parser error: test error");
    }

    #[test]
    fn test_error_from_io() {
        let io_err = std::io::Error::new(std::io::ErrorKind::NotFound, "file not found");
        let err: Error = io_err.into();
        assert!(err.to_string().contains("IO error"));
    }
}
