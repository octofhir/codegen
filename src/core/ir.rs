//! Intermediate Representation (IR) for FHIR types
//!
//! This module defines language-agnostic representations of FHIR resources,
//! datatypes, and primitives that can be transformed into any target language.

use indexmap::IndexMap;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Complete type graph representing all FHIR definitions
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct TypeGraph {
    /// Resource definitions (e.g., Patient, Observation)
    pub resources: IndexMap<String, ResourceType>,

    /// Complex datatype definitions (e.g., HumanName, Address)
    pub datatypes: IndexMap<String, DataType>,

    /// Primitive type definitions (e.g., string, boolean, date)
    pub primitives: IndexMap<String, PrimitiveType>,

    /// Profile definitions (e.g., USCorePatient)
    pub profiles: IndexMap<String, ProfileType>,

    /// FHIR version this graph represents
    pub fhir_version: FhirVersion,

    /// Metadata about generation
    pub metadata: GraphMetadata,
}

impl TypeGraph {
    /// Create a new empty type graph
    pub fn new(fhir_version: FhirVersion) -> Self {
        Self {
            resources: IndexMap::new(),
            datatypes: IndexMap::new(),
            primitives: IndexMap::new(),
            profiles: IndexMap::new(),
            fhir_version,
            metadata: GraphMetadata::default(),
        }
    }

    /// Add a resource to the graph
    pub fn add_resource(&mut self, name: String, resource: ResourceType) {
        self.resources.insert(name, resource);
    }

    /// Add a datatype to the graph
    pub fn add_datatype(&mut self, name: String, datatype: DataType) {
        self.datatypes.insert(name, datatype);
    }

    /// Add a primitive to the graph
    pub fn add_primitive(&mut self, name: String, primitive: PrimitiveType) {
        self.primitives.insert(name, primitive);
    }

    /// Add a profile to the graph
    pub fn add_profile(&mut self, name: String, profile: ProfileType) {
        self.profiles.insert(name, profile);
    }

    /// Get total number of types in graph
    pub fn total_types(&self) -> usize {
        self.resources.len() + self.datatypes.len() + self.primitives.len() + self.profiles.len()
    }
}

/// Metadata about the type graph
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct GraphMetadata {
    /// When this graph was generated
    pub generated_at: String,

    /// Generator version
    pub generator_version: String,

    /// Source packages
    pub source_packages: Vec<String>,

    /// Custom metadata
    pub custom: HashMap<String, serde_json::Value>,
}

impl Default for GraphMetadata {
    fn default() -> Self {
        Self {
            generated_at: chrono::Utc::now().to_rfc3339(),
            generator_version: env!("CARGO_PKG_VERSION").to_string(),
            source_packages: Vec::new(),
            custom: HashMap::new(),
        }
    }
}

/// FHIR version
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub enum FhirVersion {
    /// FHIR R4 (4.0.x)
    R4,
    /// FHIR R4B (4.3.x)
    R4B,
    /// FHIR R5 (5.0.x)
    R5,
    /// FHIR R6 (6.0.x)
    R6,
}

impl std::fmt::Display for FhirVersion {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            FhirVersion::R4 => write!(f, "R4"),
            FhirVersion::R4B => write!(f, "R4B"),
            FhirVersion::R5 => write!(f, "R5"),
            FhirVersion::R6 => write!(f, "R6"),
        }
    }
}

/// FHIR Resource type (e.g., Patient, Observation)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ResourceType {
    /// Resource name (e.g., "Patient")
    pub name: String,

    /// Base resource this extends (e.g., Some("DomainResource"))
    pub base: Option<String>,

    /// Properties/elements of this resource
    pub properties: Vec<Property>,

    /// Search parameters defined for this resource
    pub search_parameters: Vec<SearchParameter>,

    /// Documentation
    pub documentation: Documentation,

    /// Canonical URL
    pub url: String,

    /// Is this an abstract type?
    pub is_abstract: bool,
}

/// FHIR Complex datatype (e.g., HumanName, Address)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct DataType {
    /// Datatype name (e.g., "HumanName")
    pub name: String,

    /// Base type this extends (if any)
    pub base: Option<String>,

    /// Properties of this datatype
    pub properties: Vec<Property>,

    /// Documentation
    pub documentation: Documentation,

    /// Canonical URL
    pub url: String,

    /// Is this an abstract type?
    pub is_abstract: bool,
}

/// FHIR Primitive type (e.g., string, boolean, date)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct PrimitiveType {
    /// Primitive name (e.g., "string", "boolean")
    pub name: String,

    /// Base primitive (all primitives extend from Element)
    pub base: Option<String>,

    /// Regex pattern for validation (if any)
    pub pattern: Option<String>,

    /// Documentation
    pub documentation: Documentation,

    /// Canonical URL
    pub url: String,
}

/// FHIR Profile (constraint on a resource or datatype)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ProfileType {
    /// Profile name (e.g., "USCorePatient")
    pub name: String,

    /// Base type being constrained (e.g., "Patient")
    pub base: String,

    /// Additional constraints on properties
    pub property_constraints: Vec<PropertyConstraint>,

    /// New properties added by this profile
    pub new_properties: Vec<Property>,

    /// Documentation
    pub documentation: Documentation,

    /// Canonical URL
    pub url: String,
}

/// Property constraint in a profile
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct PropertyConstraint {
    /// Path to the property being constrained
    pub path: String,

    /// New cardinality (if changed)
    pub cardinality: Option<CardinalityRange>,

    /// New type constraints
    pub type_constraints: Vec<String>,

    /// New binding (if changed)
    pub binding: Option<ValueSetBinding>,

    /// Must support flag
    pub must_support: bool,
}

/// A property (field/element) in a resource or datatype
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Property {
    /// Property name (e.g., "name", "birthDate")
    pub name: String,

    /// Original FHIR path (e.g., "Patient.name")
    pub path: String,

    /// Type of this property
    pub property_type: PropertyType,

    /// Cardinality (min..max)
    pub cardinality: CardinalityRange,

    /// Is this a choice element? (e.g., value[x])
    pub is_choice: bool,

    /// If choice, what are the possible types?
    pub choice_types: Vec<String>,

    /// Is this a modifier element?
    pub is_modifier: bool,

    /// Is this a summary element?
    pub is_summary: bool,

    /// Terminology binding (if any)
    pub binding: Option<ValueSetBinding>,

    /// Constraints/invariants
    pub constraints: Vec<InvariantRule>,

    /// Short description
    pub short_description: String,

    /// Full definition
    pub definition: String,

    /// Comments
    pub comments: Option<String>,

    /// Example values
    pub examples: Vec<Example>,
}

/// Type of a property
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "kind")]
pub enum PropertyType {
    /// Primitive type (e.g., "string", "boolean")
    Primitive {
        /// Type name
        type_name: String,
    },

    /// Complex type (e.g., "HumanName", "Address")
    Complex {
        /// Type name
        type_name: String,
    },

    /// Reference to another resource
    Reference {
        /// Allowed target types (empty = any resource)
        target_types: Vec<String>,
    },

    /// Backbone element (inline complex type)
    BackboneElement {
        /// Properties in this backbone element
        properties: Vec<Property>,
    },

    /// Choice type (one of several types)
    Choice {
        /// Possible types for this choice
        types: Vec<String>,
    },
}

impl PropertyType {
    /// Get the primary type name (for simple cases)
    pub fn type_name(&self) -> Option<&str> {
        match self {
            PropertyType::Primitive { type_name } => Some(type_name),
            PropertyType::Complex { type_name } => Some(type_name),
            _ => None,
        }
    }
}

/// Cardinality range (min..max)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct CardinalityRange {
    /// Minimum occurrences
    pub min: u32,

    /// Maximum occurrences (None = unbounded)
    pub max: Option<u32>,
}

impl CardinalityRange {
    /// Create a required single value (1..1)
    pub fn required() -> Self {
        Self { min: 1, max: Some(1) }
    }

    /// Create an optional single value (0..1)
    pub fn optional() -> Self {
        Self { min: 0, max: Some(1) }
    }

    /// Create a required array (1..*)
    pub fn required_array() -> Self {
        Self { min: 1, max: None }
    }

    /// Create an optional array (0..*)
    pub fn optional_array() -> Self {
        Self { min: 0, max: None }
    }

    /// Is this property required?
    pub fn is_required(&self) -> bool {
        self.min >= 1
    }

    /// Is this property an array?
    pub fn is_array(&self) -> bool {
        self.max.is_none() || self.max.is_some_and(|max| max > 1)
    }

    /// Is this property optional?
    pub fn is_optional(&self) -> bool {
        self.min == 0
    }
}

/// ValueSet binding for coded elements
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ValueSetBinding {
    /// Binding strength
    pub strength: BindingStrength,

    /// ValueSet canonical URL
    pub value_set: String,

    /// Description
    pub description: Option<String>,
}

/// Binding strength
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
pub enum BindingStrength {
    /// Required (must use)
    Required,
    /// Extensible (should use)
    Extensible,
    /// Preferred (could use)
    Preferred,
    /// Example (may use)
    Example,
}

/// Invariant constraint rule
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct InvariantRule {
    /// Invariant key (e.g., "pat-1")
    pub key: String,

    /// Severity
    pub severity: ConstraintSeverity,

    /// Human description
    pub human: String,

    /// FHIRPath expression
    pub expression: Option<String>,

    /// XPath expression (legacy)
    pub xpath: Option<String>,
}

/// Constraint severity
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
pub enum ConstraintSeverity {
    /// Error (must satisfy)
    Error,
    /// Warning (should satisfy)
    Warning,
}

/// Example value
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Example {
    /// Example label
    pub label: String,

    /// Example value (as JSON)
    pub value: serde_json::Value,
}

/// Documentation for types and properties
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
pub struct Documentation {
    /// Short description
    pub short: String,

    /// Full definition
    pub definition: String,

    /// Comments
    pub comments: Option<String>,

    /// Requirements
    pub requirements: Option<String>,

    /// Usage notes
    pub usage_notes: Vec<String>,

    /// Canonical URL
    pub url: Option<String>,
}

/// Search parameter definition
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SearchParameter {
    /// Parameter name/code
    pub code: String,

    /// Parameter type
    pub param_type: SearchParamType,

    /// Description
    pub description: String,

    /// FHIRPath expression
    pub expression: Option<String>,

    /// Target resource types (for reference params)
    pub target_types: Vec<String>,
}

/// Search parameter type
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
pub enum SearchParamType {
    /// Number
    Number,
    /// Date/DateTime
    Date,
    /// String
    String,
    /// Token
    Token,
    /// Reference
    Reference,
    /// Composite
    Composite,
    /// Quantity
    Quantity,
    /// URI
    Uri,
    /// Special
    Special,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_type_graph_creation() {
        let graph = TypeGraph::new(FhirVersion::R4);
        assert_eq!(graph.fhir_version, FhirVersion::R4);
        assert_eq!(graph.total_types(), 0);
    }

    #[test]
    fn test_type_graph_add_resource() {
        let mut graph = TypeGraph::new(FhirVersion::R4);

        let resource = ResourceType {
            name: "Patient".to_string(),
            base: Some("DomainResource".to_string()),
            properties: Vec::new(),
            search_parameters: Vec::new(),
            documentation: Documentation::default(),
            url: "http://hl7.org/fhir/StructureDefinition/Patient".to_string(),
            is_abstract: false,
        };

        graph.add_resource("Patient".to_string(), resource);
        assert_eq!(graph.total_types(), 1);
        assert!(graph.resources.contains_key("Patient"));
    }

    #[test]
    fn test_cardinality_required() {
        let card = CardinalityRange::required();
        assert_eq!(card.min, 1);
        assert_eq!(card.max, Some(1));
        assert!(card.is_required());
        assert!(!card.is_array());
        assert!(!card.is_optional());
    }

    #[test]
    fn test_cardinality_optional() {
        let card = CardinalityRange::optional();
        assert_eq!(card.min, 0);
        assert_eq!(card.max, Some(1));
        assert!(!card.is_required());
        assert!(!card.is_array());
        assert!(card.is_optional());
    }

    #[test]
    fn test_cardinality_array() {
        let card = CardinalityRange::optional_array();
        assert_eq!(card.min, 0);
        assert_eq!(card.max, None);
        assert!(!card.is_required());
        assert!(card.is_array());
        assert!(card.is_optional());
    }

    #[test]
    fn test_cardinality_required_array() {
        let card = CardinalityRange::required_array();
        assert_eq!(card.min, 1);
        assert_eq!(card.max, None);
        assert!(card.is_required());
        assert!(card.is_array());
        assert!(!card.is_optional());
    }

    #[test]
    fn test_property_type_typename() {
        let prim = PropertyType::Primitive { type_name: "string".to_string() };
        assert_eq!(prim.type_name(), Some("string"));

        let complex = PropertyType::Complex { type_name: "HumanName".to_string() };
        assert_eq!(complex.type_name(), Some("HumanName"));

        let reference = PropertyType::Reference { target_types: vec!["Patient".to_string()] };
        assert_eq!(reference.type_name(), None);
    }

    #[test]
    fn test_fhir_version_display() {
        assert_eq!(FhirVersion::R4.to_string(), "R4");
        assert_eq!(FhirVersion::R4B.to_string(), "R4B");
        assert_eq!(FhirVersion::R5.to_string(), "R5");
        assert_eq!(FhirVersion::R6.to_string(), "R6");
    }

    #[test]
    fn test_graph_metadata_default() {
        let metadata = GraphMetadata::default();
        assert_eq!(metadata.generator_version, env!("CARGO_PKG_VERSION"));
        assert!(metadata.source_packages.is_empty());
        assert!(!metadata.generated_at.is_empty());
    }
}
