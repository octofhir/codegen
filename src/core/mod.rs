//! Core functionality for codegen

pub mod error;
pub mod graph_builder;
pub mod ir;
pub mod parser;
pub mod resolver;

pub use error::{Error, Result};
pub use graph_builder::TypeGraphBuilder;
pub use ir::TypeGraph;
pub use parser::StructureDefinitionParser;
pub use resolver::SchemaResolver;
