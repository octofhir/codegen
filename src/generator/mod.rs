//! Code generation traits and utilities

pub mod traits;

pub use traits::{
    CodeGenerator, FileType, GeneratedCode, GeneratedFile, GenerationManifest,
    GenerationStatistics, GeneratorCapabilities, GeneratorConfig, GeneratorMetadata,
    IdentifierContext, Language, LanguageBackend, TemplateContext, TemplateEngine,
};
