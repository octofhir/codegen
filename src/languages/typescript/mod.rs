//! TypeScript code generation

mod backend;
pub mod choice_generator;
pub mod datatype_generator;
/// Documentation generation for TypeScript with JSDoc and TypeDoc support
pub mod documentation_generator;
/// Helper methods generation for TypeScript
pub mod helpers_generator;
/// Package manifest generation for TypeScript
pub mod manifest_generator;
pub mod resource_generator;
/// Complete TypeScript SDK generation orchestrator
pub mod sdk_generator;
pub mod templates;
/// Validation functions generation for TypeScript
pub mod validation_generator;

pub use backend::TypeScriptBackend;
pub use choice_generator::ChoiceElementGenerator;
pub use datatype_generator::DatatypeGenerator;
pub use documentation_generator::DocumentationGenerator;
pub use helpers_generator::HelpersGenerator;
pub use manifest_generator::{ManifestGenerator, PackageConfig};
pub use resource_generator::ResourceGenerator;
pub use sdk_generator::TypeScriptSdkGenerator;
pub use validation_generator::ValidationGenerator;
