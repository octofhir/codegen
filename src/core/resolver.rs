//! Schema resolution via canonical-manager
//!
//! Provides type lookups and dependency resolution using the canonical manager

use crate::core::{Error, Result};
use octofhir_canonical_manager::CanonicalManager;
use octofhir_canonical_manager::resolver::ResolvedResource;
use serde_json::Value;
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::RwLock;
use tracing::{debug, warn};

/// Schema resolver using canonical-manager
pub struct SchemaResolver {
    /// Canonical manager instance
    manager: Arc<CanonicalManager>,

    /// Cache of resolved StructureDefinitions
    cache: Arc<RwLock<HashMap<String, Value>>>,

    /// Cache of primitive type checks
    primitives: Arc<RwLock<HashMap<String, bool>>>,
}

impl SchemaResolver {
    /// Create a new schema resolver
    ///
    /// # Arguments
    ///
    /// * `manager` - Canonical manager instance
    pub fn new(manager: Arc<CanonicalManager>) -> Self {
        Self {
            manager,
            cache: Arc::new(RwLock::new(HashMap::new())),
            primitives: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    /// Resolve a type by canonical URL
    ///
    /// # Arguments
    ///
    /// * `canonical_url` - Canonical URL of the type to resolve
    ///
    /// # Returns
    ///
    /// StructureDefinition JSON if found
    pub async fn resolve_type(&self, canonical_url: &str) -> Result<Option<Value>> {
        // Check cache first
        {
            let cache = self.cache.read().await;
            if let Some(cached) = cache.get(canonical_url) {
                debug!("Cache hit for {}", canonical_url);
                return Ok(Some(cached.clone()));
            }
        }

        // Resolve via canonical manager
        debug!("Resolving {} via canonical-manager", canonical_url);

        match self.manager.resolve(canonical_url).await {
            Ok(resolved) => {
                let content = resolved.resource.content.clone();

                // Cache the result
                {
                    let mut cache = self.cache.write().await;
                    cache.insert(canonical_url.to_string(), content.clone());
                }

                Ok(Some(content))
            }
            Err(e) => {
                warn!("Failed to resolve {}: {}", canonical_url, e);
                Ok(None)
            }
        }
    }

    /// Resolve base type for a StructureDefinition
    ///
    /// # Arguments
    ///
    /// * `base_url` - Canonical URL of the base type
    ///
    /// # Returns
    ///
    /// Base type StructureDefinition if found
    pub async fn resolve_base_type(&self, base_url: &str) -> Result<Option<Value>> {
        self.resolve_type(base_url).await
    }

    /// Check if a type is a FHIR primitive
    ///
    /// Uses canonical manager to look up the type and check its kind
    pub async fn is_primitive_type(&self, type_code: &str) -> Result<bool> {
        // Check cache first
        {
            let cache = self.primitives.read().await;
            if let Some(&is_prim) = cache.get(type_code) {
                return Ok(is_prim);
            }
        }

        // Build canonical URL for core FHIR types
        let canonical_url = format!("http://hl7.org/fhir/StructureDefinition/{}", type_code);

        if let Some(sd) = self.resolve_type(&canonical_url).await? {
            let kind = sd.get("kind").and_then(|v| v.as_str()).unwrap_or("complex-type");

            let is_prim = kind == "primitive-type";

            // Cache the result
            {
                let mut cache = self.primitives.write().await;
                cache.insert(type_code.to_string(), is_prim);
            }

            Ok(is_prim)
        } else {
            // If not found, assume it's a complex type
            Ok(false)
        }
    }

    /// Resolve a value set by canonical URL
    ///
    /// # Arguments
    ///
    /// * `value_set_url` - Canonical URL of the value set
    ///
    /// # Returns
    ///
    /// ValueSet resource if found
    pub async fn resolve_value_set(&self, value_set_url: &str) -> Result<Option<Value>> {
        debug!("Resolving value set: {}", value_set_url);

        match self.manager.resolve(value_set_url).await {
            Ok(resolved) => Ok(Some(resolved.resource.content)),
            Err(e) => {
                warn!("Failed to resolve value set {}: {}", value_set_url, e);
                Ok(None)
            }
        }
    }

    /// Get all resources of a specific type
    ///
    /// # Arguments
    ///
    /// * `resource_type` - FHIR resource type (e.g., "StructureDefinition")
    /// * `limit` - Maximum number of resources to return (defaults to 1000)
    ///
    /// # Returns
    ///
    /// List of matching resources
    pub async fn get_resources_by_type(
        &self,
        resource_type: &str,
        limit: Option<usize>,
    ) -> Result<Vec<ResolvedResource>> {
        let limit = limit.unwrap_or(1000);
        debug!("Searching for resources of type: {} (limit: {})", resource_type, limit);

        let query = self
            .manager
            .search()
            .await
            .resource_type(resource_type)
            .limit(limit)
            .execute()
            .await
            .map_err(|e| Error::CanonicalManager(e.to_string()))?;

        // Convert search results to resolved resources
        let mut resources = Vec::new();
        for result in query.resources {
            // Resolve each resource to get full content
            if let Some(url) = &result.resource.url
                && let Ok(resolved) = self.manager.resolve(url).await
            {
                resources.push(resolved);
            }
        }

        Ok(resources)
    }

    /// Get all StructureDefinitions for a FHIR version
    ///
    /// # Arguments
    ///
    /// * `fhir_version` - FHIR version (e.g., "R4", "R5")
    ///
    /// # Returns
    ///
    /// List of StructureDefinition resources
    pub async fn get_all_structure_definitions(&self, _fhir_version: &str) -> Result<Vec<Value>> {
        debug!("Loading all StructureDefinitions");

        // Use a high limit to get all StructureDefinitions (typically 145+ for R4)
        let resources = self.get_resources_by_type("StructureDefinition", Some(1000)).await?;

        Ok(resources.into_iter().map(|r| r.resource.content).collect())
    }

    /// Clear all caches
    pub async fn clear_cache(&self) {
        let mut cache = self.cache.write().await;
        cache.clear();

        let mut prim_cache = self.primitives.write().await;
        prim_cache.clear();

        debug!("Schema resolver cache cleared");
    }

    /// Get cache statistics
    pub async fn cache_stats(&self) -> CacheStats {
        let cache = self.cache.read().await;
        let primitives = self.primitives.read().await;

        CacheStats { type_cache_size: cache.len(), primitive_cache_size: primitives.len() }
    }
}

/// Cache statistics
#[derive(Debug, Clone)]
pub struct CacheStats {
    /// Number of cached types
    pub type_cache_size: usize,
    /// Number of cached primitive type checks
    pub primitive_cache_size: usize,
}

#[cfg(test)]
mod tests {
    use super::*;
    use octofhir_canonical_manager::FcmConfig;
    use tempfile::TempDir;

    async fn create_test_manager() -> Arc<CanonicalManager> {
        let temp_dir = TempDir::new().unwrap();
        let mut config = FcmConfig::default();
        config.storage.packages_dir = temp_dir.path().join("packages");
        config.storage.cache_dir = temp_dir.path().join("cache");
        config.storage.index_dir = temp_dir.path().join("index");
        Arc::new(CanonicalManager::new(config).await.unwrap())
    }

    #[tokio::test]
    async fn test_resolver_creation() {
        let manager = create_test_manager().await;
        let resolver = SchemaResolver::new(manager);

        let stats = resolver.cache_stats().await;
        assert_eq!(stats.type_cache_size, 0);
        assert_eq!(stats.primitive_cache_size, 0);
    }

    #[tokio::test]
    async fn test_cache_clear() {
        let manager = create_test_manager().await;
        let resolver = SchemaResolver::new(manager);

        resolver.clear_cache().await;

        let stats = resolver.cache_stats().await;
        assert_eq!(stats.type_cache_size, 0);
    }

    #[tokio::test]
    async fn test_resolve_nonexistent_type() {
        let manager = create_test_manager().await;
        let resolver = SchemaResolver::new(manager);

        let result = resolver.resolve_type("http://example.com/nonexistent").await;
        assert!(result.is_ok());
        assert!(result.unwrap().is_none());
    }

    #[tokio::test]
    async fn test_cache_stats() {
        let manager = create_test_manager().await;
        let resolver = SchemaResolver::new(manager);

        let stats = resolver.cache_stats().await;
        assert_eq!(stats.type_cache_size, 0);
        assert_eq!(stats.primitive_cache_size, 0);
    }
}
