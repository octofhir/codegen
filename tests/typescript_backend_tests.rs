use insta::assert_snapshot;
use octofhir_codegen::core::ir::*;
use octofhir_codegen::generator::LanguageBackend;
use octofhir_codegen::languages::typescript::TypeScriptBackend;

#[test]
fn test_complex_interface_generation() {
    let backend = TypeScriptBackend::new();

    let doc = Documentation {
        short: "Patient demographics and information".to_string(),
        definition: "Demographics and other administrative information about an individual or animal receiving care or other health-related services.".to_string(),
        url: Some("http://hl7.org/fhir/StructureDefinition/Patient".to_string()),
        ..Default::default()
    };

    let props = vec![
        ("resourceType".to_string(), "\"Patient\"".to_string(), false),
        ("id".to_string(), "id".to_string(), true),
        ("active".to_string(), "boolean".to_string(), true),
        ("name".to_string(), "HumanName[]".to_string(), true),
    ];

    let output = backend.generate_interface("Patient", Some("DomainResource"), &props, Some(&doc));

    assert_snapshot!("patient_interface", output);
}

#[test]
fn test_type_alias_generation() {
    let backend = TypeScriptBackend::new();

    let doc = Documentation {
        short: "Patient gender".to_string(),
        definition: "Administrative gender - male | female | other | unknown".to_string(),
        ..Default::default()
    };

    let output = backend.generate_type_alias(
        "PatientGender",
        "\"male\" | \"female\" | \"other\" | \"unknown\"",
        Some(&doc),
    );

    assert_snapshot!("patient_gender_type", output);
}

#[test]
fn test_primitive_type_mapping() {
    let backend = TypeScriptBackend::new();

    let boolean_type = PropertyType::Primitive { type_name: "boolean".to_string() };
    assert_eq!(backend.map_type(&boolean_type), "boolean");

    let integer_type = PropertyType::Primitive { type_name: "integer".to_string() };
    assert_eq!(backend.map_type(&integer_type), "number");

    let string_type = PropertyType::Primitive { type_name: "string".to_string() };
    assert_eq!(backend.map_type(&string_type), "string");

    let datetime_type = PropertyType::Primitive { type_name: "dateTime".to_string() };
    assert_eq!(backend.map_type(&datetime_type), "string");
}

#[test]
fn test_complex_type_mapping() {
    let backend = TypeScriptBackend::new();

    let complex_type = PropertyType::Complex { type_name: "HumanName".to_string() };
    assert_eq!(backend.map_type(&complex_type), "HumanName");
}

#[test]
fn test_reference_type_mapping() {
    let backend = TypeScriptBackend::new();

    // Empty reference
    let empty_ref = PropertyType::Reference { target_types: vec![] };
    assert_eq!(backend.map_type(&empty_ref), "Reference");

    // Single target
    let single_ref = PropertyType::Reference { target_types: vec!["Patient".to_string()] };
    assert_eq!(backend.map_type(&single_ref), "Reference<Patient>");

    // Multiple targets
    let multi_ref = PropertyType::Reference {
        target_types: vec!["Patient".to_string(), "Organization".to_string()],
    };
    assert_eq!(backend.map_type(&multi_ref), "Reference<Patient | Organization>");
}

#[test]
fn test_choice_type_mapping() {
    let backend = TypeScriptBackend::new();

    let choice_type = PropertyType::Choice {
        types: vec!["string".to_string(), "number".to_string(), "boolean".to_string()],
    };
    assert_eq!(backend.map_type(&choice_type), "string | number | boolean");
}

#[test]
fn test_generate_imports() {
    let backend = TypeScriptBackend::new();

    // Empty dependencies
    let empty_imports = backend.generate_imports(&[]);
    assert_eq!(empty_imports.len(), 0);

    // Single dependency
    let single_imports = backend.generate_imports(&["HumanName".to_string()]);
    assert_eq!(single_imports.len(), 1);
    assert_eq!(single_imports[0], "import { HumanName } from './types';");

    // Multiple dependencies
    let multi_imports = backend.generate_imports(&["HumanName".to_string(), "Address".to_string()]);
    assert_eq!(multi_imports.len(), 1);
    assert_eq!(multi_imports[0], "import { HumanName, Address } from './types';");
}

#[test]
fn test_doc_comment_generation() {
    let backend = TypeScriptBackend::new();

    let doc = Documentation {
        short: "Patient resource".to_string(),
        definition:
            "This is a longer definition that describes the patient resource in more detail."
                .to_string(),
        comments: Some("Additional comments here".to_string()),
        url: Some("http://hl7.org/fhir/StructureDefinition/Patient".to_string()),
        ..Default::default()
    };

    let output = backend.generate_doc_comment(&doc);

    assert_snapshot!("doc_comment", output.join("\n"));
}

#[test]
fn test_interface_without_doc() {
    let backend = TypeScriptBackend::new();

    let props = vec![
        ("id".to_string(), "string".to_string(), true),
        ("name".to_string(), "string".to_string(), false),
    ];

    let output = backend.generate_interface("SimpleInterface", None, &props, None);

    assert_snapshot!("simple_interface", output);
}

#[test]
fn test_type_alias_without_doc() {
    let backend = TypeScriptBackend::new();

    let output = backend.generate_type_alias("Status", "\"active\" | \"inactive\"", None);

    assert_snapshot!("simple_type_alias", output);
}
