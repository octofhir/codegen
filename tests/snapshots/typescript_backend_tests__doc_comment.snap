---
source: tests/typescript_backend_tests.rs
expression: "output.join(\"\\n\")"
---
/**
 * Patient resource
 *
 * This is a longer definition that describes the patient resource in more detail.
 *
 * Additional comments here
 *
 * @see http://hl7.org/fhir/StructureDefinition/Patient
 */
